Source/ParleyTests/MessagesManagerTests.swift

```diff
@@ -0,0 +1,252 @@
+        XCTAssertEqual(indexPaths[0].row, 3, "first index path should insert at index 3.")
+        XCTAssertEqual(messagesManager.messages[4].type, .agentTyping, "Fifth message should the typing indicator.")
+        XCTAssertEqual(messagesManager.messages.last, messagesManager.messages[4], "Tying indicator should be the last message.")
+    }
+    
+    func testRetrievingOldestMessage() {
+        XCTAssertNil(messagesManager.getOldestMessage())
+        
+        setWelcomeMessage()
+        messagesManager.loadCachedData()
+        XCTAssertNil(messagesManager.getOldestMessage())
+        
+        let firstMessage = createUserMessage("Good morning")
+        _ = messagesManager.add(firstMessage)
+        XCTAssertEqual(firstMessage, messagesManager.getOldestMessage())
+        
+        let secondMessage = createUserMessage("Second message")
+        _ = messagesManager.add(secondMessage)
+        XCTAssertEqual(firstMessage, messagesManager.getOldestMessage())
+    }
+    
+    func testUpdateMessage() {
+        let originalMessage = Message()
+        originalMessage.type = .user
+        originalMessage.uuid = UUID().uuidString
+        let originalMessageText = "First original message"
+        originalMessage.message = originalMessageText
+        
+        _ = messagesManager.add(originalMessage)
+        // A date message gets added, last message is the user's message.
+        XCTAssertEqual(messagesManager.messages.last?.uuid, originalMessage.uuid)
+        
+        let updatedMessage = Message()
+        updatedMessage.type = .user
+        updatedMessage.uuid = originalMessage.uuid
+        let updatedMessageText = "Updated message text"
+        updatedMessage.message = updatedMessageText
+        _ = messagesManager.update(updatedMessage)
+        
+        XCTAssertNotEqual(messagesManager.messages.last!.message, originalMessageText)
+        XCTAssertEqual(messagesManager.messages.last!.message, updatedMessageText)
+        XCTAssertEqual(messagesManager.messages.last!.uuid, originalMessage.uuid)
+    }
+}
+
+internal extension MessagesManagerTests {
+    
+    @discardableResult
+    func addInfoMessageAndFirstUserMessage() -> [IndexPath] {
+        setWelcomeMessage()
+        messagesManager.loadCachedData()
+        
+        let firstMessage = createUserMessage("Good morning")
+        return messagesManager.add(firstMessage)
+    }
+    
+    func setWelcomeMessage() {
+        XCTAssertTrue(dataSource.set(MESSAGE_WELCOME_TEXT, forKey: kParleyCacheKeyMessageInfo))
+    }
+    
+    func createUserMessage(_ message: String, date: Date = Date()) -> Message {
+        let userMessage = Message()
+        userMessage.message = message
+        userMessage.time = date
+        userMessage.type = .user
+        return userMessage
+    }
+    
+    func getYesterdayDate() -> Date? {
+        let today = Date()
+        guard let yesterday = Calendar.current.date(byAdding: .day, value: -1, to: today) else { return nil }
```

Comment:
Better to throw instead of `nil`, yesterday should in no case be unknown in these tests

Reply:
Perfect! ✅